### PR TITLE
Peer reviewed article subscribe to discussion updates

### PR DESCRIPTION
**Pull request by laurensmartina**

### CODE REVIEW

**Comment by tammoterhark** [comment]
image

Moet zijn: "Houd me op de hoogte."

File: system/datatypes/peer_reviewed_article.php [hunk 1]
```diff
@@ -718,19 +735,85 @@ public function commentConfirmAction(HyphaRequest $request) {
 			return null;
 		}
 
-		$pending = (bool)$comment->getAttribute(self::FIELD_NAME_DISCUSSION_COMMENT_PENDING);
-		if ($pending) {
+		$pending = [];
```
Comment (matthijskooijman) [inline system/datatypes/peer_reviewed_article.php]: I think pending is no longer a good variable name here. Something like `$confirmed` is better, since it indicates that something was confirmed (even though that does mean it used to be pending). Also, I think that simply splitting this into two variables (e.g. `$confirmedComment` and `$confirmedSubscription`) is more clear.

Finally: Is there ever a case where you can confirm a subscription without confirming a comment? Probably not in practice, but maybe it is enough to track this separately just for abnormal circumstances?

File: system/datatypes/peer_reviewed_article.php [hunk 2]
```diff
@@ -1233,14 +1363,53 @@ private function sendNewCommentMail(HyphaDomElement $comment) {
...
+			'title' => $title
 		]);
+		$subscriberMessage = __('art-comment-subscriber-body', $messageArgs);
+
+		// also send newly create comment to all who subscribed to this discussion, with unsubscribe link
+		foreach ($this->getSubscribers($comment) as $subscriber) {
+			$email = $subscriber->getAttribute(self::FIELD_NAME_DISCUSSION_SUBSCRIBER_EMAIL);
+			$code = $subscriber->getAttribute(self::FIELD_NAME_DISCUSSION_SUBSCRIBER_UNSUBSCRIBE_CODE);
+			// include email for server record purposes
+			$path = hypha_substitute(self::PATH_UNSUBSCRIBE_CODE, ['address' => htmlspecialchars($email), 'code' => htmlspecialchars($code)]);
```
Comment (matthijskooijman) [inline system/datatypes/peer_reviewed_article.php]: I think this should not do `htmlspecialchars`, if anything it should do `rawurlencode` (since you're building a URL here. The `htmlspecialchars` should be done below, on the full `$linkToUnsubscribe` when it is inserted into the `href` attribute.

File: system/datatypes/peer_reviewed_article.php [hunk 3]
```diff
@@ -1233,14 +1363,53 @@ private function sendNewCommentMail(HyphaDomElement $comment) {
...
+			$linkToUnsubscribe = $this->constructFullPath($this->pagename . '/' . $path);
+			$message = $subscriberMessage;
+			$message .= '<p><a href="' . $linkToUnsubscribe . '">' . htmlspecialchars(__('art-unsubscribe')) . '</a></p>';
```
Comment (matthijskooijman) [inline system/datatypes/peer_reviewed_article.php]: I wonder if this unsubsribe stuff should be appended, or maybe included in the `__('art-comment-subscriber-body')` string?

File: system/datatypes/peer_reviewed_article.php [hunk 4]
```diff
@@ -1117,6 +1243,10 @@ private function createCommentForm($type, array $values = [], HyphaDomElement $d
 			$vars['un-anonymous'] = __('art-comment-unanonymous');
 			$vars['email'] = __('art-email');
 			$vars['field-name-email'] = self::FIELD_NAME_DISCUSSION_COMMENTER_EMAIL . $dataFieldSuffix;
+			$vars['subscribe'] = __('art-subscribe');
+			$vars['info-button-subscribe'] = makeInfoButton('help-art-subscribe');
+			$vars['field-name-subscribe'] = self::FIELD_NAME_DISCUSSION_COMMENTER_SUBSCRIBE . $dataFieldSuffix;
+			$vars['field-value-subscribe-me'] = self::FIELD_VALUE_DISCUSSION_COMMENTER_SUBSCRIBE_ME;
```
Comment (matthijskooijman) [inline system/datatypes/peer_reviewed_article.php]: Why so much code over the checkbox value here? Just any non-empty value would be sufficient, and then in the handling just check to see if the value is set?

There is also some code to pass a value for the checkbox as part of `$formData`, which just sets the same value again, which also seems pointless to me? Why is all this needed?

File: system/datatypes/peer_reviewed_article.php [hunk 5]
```diff
@@ -1108,6 +1231,9 @@ private function createCommentForm($type, array $values = [], HyphaDomElement $d
 			<div class="section" style="padding:5px; margin-bottom:5px; position:relative;">
 				<strong><label for="[[field-name-email]]">[[email]]</label></strong> <input type="text" id="[[field-name-email]]" name="[[field-name-email]]" />
 			</div>
+			<div class="section" style="padding:5px; margin-bottom:5px; position:relative;">
+				<strong><label for="[[field-name-subscribe]]">[[subscribe]]</label></strong> <input value="[[field-value-subscribe-me]]" type="checkbox" id="[[field-name-subscribe]]" name="[[field-name-subscribe]]" /> [[info-button-subscribe]]
+			</div>
```
Comment (matthijskooijman) [inline system/datatypes/peer_reviewed_article.php]: I haven't checked, but this looks like the subscribe checkbox is *always* shown, even for logged in users that are automatically subscribed currently (and do not need to enter or confirm their e-mail). Or am I mistaken?

File: system/languages/en.php [hunk 6]
```diff
@@ -410,6 +418,7 @@
 		"help-art-excerpt" => "In this field you can type a short excerpt of your article.",
 		"help-art-method" => "Please type a bio of yourself containing relevant experience or possible biasses. \n If relevant, also let the reader know how the article came into existence.",
 		"help-art-text" => "Type your article here or paste from a text editor. \n If you want to paste from Word, Pages or other word processors, \n please use the 'paste from word' button. \nImages should not be pasted but \nuploaded using the 'image' button.",
+		"help-art-subscribe" => "Keep informed about this discussion, by receiving an email with every new comment.",
```
Comment (matthijskooijman) [inline system/languages/en.php]: Should this be "Keep *me* informed..."?

File: system/languages/nl.php [hunk 7]
```diff
@@ -335,9 +335,11 @@
...
+		"art-confirm-comment-body" => "Op de website van [[sitename]] heeft u (of iemand die uw emailadres gebruikte) net een reactie geplaatst bij het artikel: [[title]].<br /><br /> ------------------------------ <br />De reactie was:<br />[[comment]]<br />------------------------------<br />[[sitename]] is een platform voor constructieve discussie, daarom hebben we wat regels opgesteld voor commentatoren. Overweeg of uw commentaar nieuwe informatie of overwegingen aan de discussie toevoegt. Houd de toon beschaafd. Vermeld uw volledige naam. De redactie behoudt zich het recht voor om uw identiteit bekend te maken als u dat zelf niet doet. De redactie kan uw commentaar verwijderen als het niet aan de criteria voldoet.<br/>Als u wilt dat deze reactie geplaatst wordt, <b>bevestig dit dan door op onderstaande link te klikken</b>. Deze actie is nodig om spam te voorkomen.<br /><a href=\"[[link]]\">[[link]]</a><br /><br/>Dit is een geautomatiseerde mail. Antwoorden hierop heeft geen zin.",
```
Comment (matthijskooijman) [inline system/languages/nl.php]: This seems like an unrelated change?

File: system/languages/en.php (hunk 6 above)
Comment (matthijskooijman) [reply system/languages/en.php]: Or maybe "Stay informed..."?

**Comment by laurensmartina** [comment]
> Moet zijn: "Houd me op de hoogte."

fixed

File: system/datatypes/peer_reviewed_article.php [hunk 8]
```diff
@@ -541,31 +533,24 @@ public function discussionAction(HyphaRequest $request) {
...
-			$this->sendCommentConfirmMail($comment);
-		}
-		if (!$review && !$pending) {
-			$this->sendNewCommentMail($comment);
-		}
-
-		if ($pending) {
-			notify('success', ucfirst(__('art-comment-pending')));
+		if ($review) {
+			$blocking ? $this->sendBlockingDiscussionMail($discussion) : null;
```
[the remark this answers is not in the document]
Comment (laurensmartina) [reply system/datatypes/peer_reviewed_article.php]: fixed

File: system/datatypes/peer_reviewed_article.php (hunk 1 above)
Comment (laurensmartina) [reply system/datatypes/peer_reviewed_article.php]: fixed

File: system/datatypes/peer_reviewed_article.php (hunk 3 above)
Comment (laurensmartina) [reply system/datatypes/peer_reviewed_article.php]: Don't want to risk it being excluded.

File: system/datatypes/peer_reviewed_article.php (hunk 2 above)
Comment (laurensmartina) [reply system/datatypes/peer_reviewed_article.php]: fixed

File: system/datatypes/peer_reviewed_article.php (hunk 4 above)
Comment (laurensmartina) [reply system/datatypes/peer_reviewed_article.php]: Good to know. And a good indication that the workings of the HTMLForm is obscure.
Removed redundant code.

File: system/datatypes/peer_reviewed_article.php (hunk 5 above)
Comment (laurensmartina) [reply system/datatypes/peer_reviewed_article.php]: This part of the code is only used for non-users.

File: system/languages/en.php (hunk 6 above)
Comment (laurensmartina) [reply system/languages/en.php]: fixed

File: system/languages/nl.php (hunk 7 above)
Comment (laurensmartina) [reply system/languages/nl.php]: moved from this commit

**Comment by matthijskooijman** [comment]
I've also rebased on top of master.

**Comment by laurensmartina** [comment]
@matthijskooijman I live the changes you made, and rebased the fixup

**Comment by tammoterhark** [comment]
I think we will have to find out if the functionality is sufficiently clear. We'll see.

**Comment by matthijskooijman** [comment]
@laurensmartina  what does "I live the changes you made"  mean? Did you mean "reviewed"?

Regardless, shall we merge this, put it in test.destadsbron.nl and include it in the upcoming deployment?

**Comment by laurensmartina** [comment]
@matthijskooijman , it was a typo.. `live` should have been `like`.
